app/models/visitors_step.rb

```diff
@@ -1,18 +1,65 @@
 class VisitorsStep
   include NonPersistedModel
-  include Person
+
+  class Visitor
+    include NonPersistedModel
+    include Person
+
+    attribute :first_name, String
+    attribute :last_name, String
+    attribute :date_of_birth, Date
+  end
 
   attribute :email_address, String
+  attribute :phone_no, String
   attribute :override_delivery_error, Boolean
   attribute :delivery_error_type, String
   attribute :delivery_error_occurred, Boolean
-  attribute :phone_no, String
+  attribute :visitors, Array[Visitor]
 
   validates :email_address, presence: true
   validates :phone_no, presence: true, length: { minimum: 9 }
 
   validate :validate_email
 
+  # Return at least Prison::MAX_VISITORS visitors, filling with new instances
+  # as needed. The regular #visitors method will return only those visitors
+  # actually supplied via filled fields (or one blank primary visitor).
+  def backfilled_visitors
+    existing = visitors
+    num_needed = Prison::MAX_VISITORS - existing.count
+    backfill = num_needed.times.map { Visitor.new }
+    existing + backfill
+  end
+
+  def visitors_attributes=(params)
+    # params is of the form
+    # {"0" => {"foo" => "bar"}, "1" => {"foo" => "baz"}}
+    # so we sort by key and take the values. We throw away empty visitors.
+    pruned = ParameterPruner.new.prune(
+      params.sort_by { |k, _| k.to_i }.map(&:last)
+    )
+
+    # We always want at least one visitor. Leaving the rest blank is fine, but
+    # the first one must both exist and be valid.
+    self.visitors = pruned.empty? ? [{}] : pruned.take(Prison::MAX_VISITORS)
+  end
+
+  def valid?(*)
+    # This must be eager because we want to show errors on all objects.
+    results = [super]
```

Comment:
Yes, that's a bit terser, isn't it? (I'm going to use `inject` instead of `reduce` for consistency, but that's not significant.)